src/script/printer/tir/type.cc

```diff
@@ -18,39 +18,45 @@
  */
 #include <tvm/ir/type.h>
 
+#include "../util.h"
 #include "./utils.h"
 
 namespace tvm {
 namespace script {
 namespace printer {
 
 TVM_STATIC_IR_FUNCTOR(IRDocsifier, vtable)
-    .set_dispatch<PrimType>("tir", [](PrimType ty, IRDocsifier p) -> Doc {
-      using runtime::DLDataType2String;
-      return TIR(p)->Attr(DLDataType2String(ty->dtype));
+    .set_dispatch<PrimType>("tir", [](PrimType raw_ty, ObjectPath path, IRDocsifier p) -> Doc {
+      auto ty = MakeTraced(raw_ty, path);
+      auto dtype = ty.GetAttr<DataType>("dtype");
```

Comment:
I am thinking about how to get more compile-time safety on attribute access. The current way would work but it's brittle to refactoring and providing less visibility when people want to track where an attribute gets accessed.

Have you consider using a macro-based approach? For example,
```
TRACE(ty, ty->dtype)
// Expands to
GetTracedObject(ty->dtype, ty, "ty->dtype");
// Then the string "ty->dtype" is split inside the function to get the attr path
```

Reply:
Yes, I thought about using macros as well, and I think it's a viable solution. One annoying problem is that some C++ member names don't match the attribute names defined in VisitAttrs (for example, `Var::name_hint` is listed as "name"), so the code becomes brittle in another way.

We could mitigate the lack of compile-time safety of the current approach with test coverage -- if someone changes an attribute name or type, the test will definitely fail. (With the exception of someone changing the type from one enum to another -- we won't detect that because of the limitations of AttrVisitor's design).

That said, I don't have a strong opinion and I don't mind changing the code. Just let me know which option you prefer.

By the way, the macro approach I'd go with would look something like
```
#define GET_ATTR_TRACED(traced_object, attr_name) \
    MakeTraced((traced_object).Get()->attr_name, (traced_object).GetPath().Attr(#attr_name)

TracedObject<tir::LetStmt> stmt = ...;
TracedObject<tir::Stmt> body = GET_ATTR_TRACED(stmt, body);
```

Reply:
Yet another option we could try is to use pointers-to-member, i.e. something like
```
auto body = stmt.GetAttr(&tir::LetStmt::body);
```
Then we can use reflection to figure out the attribute name to store in the path, or alternatively just use member pointers instead of names in the paths, but then the paths won't be nicely printable.

Reply:
> One annoying problem is that some C++ member names don't match the attribute names defined in VisitAttrs (for example, Var::name_hint is listed as "name"), so the code becomes brittle in another way.

This is a good point. Let me think a little bit longer on which side we want to choose on this tradeoff. I will take a look on how many exceptions (like name_hint -> name) are there and whether we can simply update the code to align the visited name and the real c++ attribute name. IMHO the tooling support (go to definition, auto complete, find usage, error detection in editor) is quite important and it will be a difficult problem if we do attr access in a pure reflection-based approach.

Reply:
I'm with you. I'm leaning towards the member-pointer option: it doesn't rely on macros and doesn't suffer from the attribute name mismatch problem. What do you think? I tried it and it works:

```
  template <typename T>
  typename detail::TracedObjectWrapperSelector<T>::Type GetAttr_New(T ObjectType::* member_ptr) const {
    using WrapperType = typename detail::TracedObjectWrapperSelector<T>::Type;
    const ObjectType* node = static_cast<const ObjectType*>(ref_.get());
    const T& attr = node->*member_ptr;
    const char* attr_key = ICHECK_NOTNULL(detail::GetAttrKeyByAddress(node, &attr));
    return WrapperType(attr, path_->Attr(attr_key));
  }

```

Reply:
This looks good! Even though it requires more typing in the code, it avoids using macro and the potential inconsistent attr path in VisitAttr.